feat(provisioner): propagate trace info

coryb opened the pull request:
If tracing is enabled, propagate the trace information to the terraform provisioner via environment variables.  This sets the `TRACEPARENT` environment variable using the default W3C trace propagators.  Users can choose to continue the trace by adding new spans in the provisioner by reading from the environment like:

	ctx := env.ContextWithRemoteSpanContext(context.Background(), os.Environ())

coryb committed:
feat(provisioner): propagate trace info

If tracing is enabled, propagate the trace information to the terraform
provisioner via environment variables.  This sets the `TRACEPARENT`
environment variable using the default W3C trace propagators.  Users
can choose to continue the trace by adding new spans in the provisioner
by reading from the environment like:

	ctx := env.ContextWithRemoteSpanContext(context.Background(), os.Environ())

spikecurtis commented:
Thanks for the PR!

Can you say a little more about the use case for this change? It appears that `terraform` itself [doesn't yet process TRACEPARENT](https://github.com/hashicorp/terraform/issues/35444). Are there provisioners you use that do process it?

coryb commented:
> Can you say a little more about the use case for this change? It appears that terraform itself https://github.com/hashicorp/terraform/issues/35444. Are there provisioners you use that do process it?

Correct, terraform itself won't use it, but it will maintain it in the environment for the provider binaries it calls.  We (Netflix) have a custom provider to implement our deployment strategies and would like the ability to continue traces that were started in Coder or further upstream.  For example, our provider Create and Update methods will just call:

```
ctx := env.ContextWithRemoteSpanContext(context.Background(), os.Environ())
```
Then any new spans created will use the inherited span context.

The primary use case is to enhance our integration testing.  In our tests, we start a trace and can see the spans between our test into Coder where it invokes terraform.  Currently, in our terraform provider we have to start a new trace, where we collect spans through to workspace startup.   With this patch, we will be able to connect our traces from the test entirely through to workspace startup, which should help with visibility/debugging.

spikecurtis commented:
That makes sense!

I have to say, from a supply-chain security standpoint, we are very reluctant to add a dependency on `github.com/coryb/otelbundle` if we can avoid it. If you're OK with it, we'd be willing to accept the `propagation/env` implementation as a contribution directly to this repo.

The other thing I'd ask for is a unit test of the propagation.

coryb commented:
> I have to say, from a supply-chain security standpoint, we are very reluctant to add a dependency on github.com/coryb/otelbundle if we can avoid it. If you're OK with it, we'd be willing to accept the propagation/env implementation as a contribution directly to this repo.

Thanks, works for me, I will update.  You should be able to swap out the implementation with an official OTEL support once they get that sorted.

coryb commented:
@spikecurtis copied the required code to inject the trace env, and made all the functions unexported.  Also added a test to ensure the TRACEPARENT env var is set/updated.

spikecurtis commented on provisioner/terraform/otelenv.go:
```diff
@@ -0,0 +1,74 @@
+			// don't directly update the slice so we don't modify the slice
+			// passed in
+			newEnv := slices.Clone(c.Env)
+			newEnv = append(newEnv[:i], append([]string{fmt.Sprintf("%s=%s", key, value)}, newEnv[i+1:]...)...)
```
Once you've cloned the initial slice, seems like it would be simpler to just modify in place, rather than re-slice and append.

coryb replied:
hah, I think that code went through too many refactors and came out nonsense. I have simplified it.

spikecurtis commented on provisioner/terraform/otelenv.go:
```diff
@@ -0,0 +1,74 @@
+// will never be modified.
+func otelEnvInject(ctx context.Context, environ []string) []string {
+	c := &envCarrier{Env: environ}
+	otel.GetTextMapPropagator().Inject(ctx, c)
```
Would be more comfortable with catching any panics and returning an error here. `Inject` sure sounds like it will only need `Set`, but it's an interface that takes the `TextMapCarrier` interface, so we don't know for sure that it will _never_ call `Keys` or `Get`.

coryb replied:
I went ahead and just implemented Keys/Get, they were trivial.  Added trivial tests to ensure they work.  We should not need them, but should future proof things just in case.

spikecurtis commented on provisioner/terraform/otelenv_test.go:
```diff
@@ -0,0 +1,63 @@
+
+var _ sdktrace.IDGenerator = (*testIDGenerator)(nil)
+
+func (testIDGenerator) NewIDs(ctx context.Context) (trace.TraceID, trace.SpanID) {
```
```suggestion
func (testIDGenerator) NewIDs(_ context.Context) (trace.TraceID, trace.SpanID) {
```

spikecurtis commented on provisioner/terraform/otelenv_test.go:
```diff
@@ -0,0 +1,63 @@
+	return traceID, spanID
+}
+
+func (testIDGenerator) NewSpanID(ctx context.Context, traceID trace.TraceID) trace.SpanID {
```
```suggestion
func (testIDGenerator) NewSpanID(_ context.Context, _ trace.TraceID) trace.SpanID {
```

spikecurtis commented on provisioner/terraform/otelenv.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+// TODO: replace this with the upstream OTEL env propagation when it is
+// released.
```
👍

spikecurtis commented on provisioner/terraform/otelenv_test.go:
```diff
@@ -0,0 +1,63 @@
+package terraform // nolint:testpackage
```
file should be named `otelenv_internal_test.go` then you don't need this `nolint`

spikecurtis reviewed (commented):
Few nits inline, but I think this is very close.

coryb committed:
Update provisioner/terraform/otelenv_test.go

Co-authored-by: Spike Curtis <spike@spikecurtis.com>

coryb committed:
Update provisioner/terraform/otelenv_test.go

Co-authored-by: Spike Curtis <spike@spikecurtis.com>

coryb committed:
update for feedback, remove unimplemented panics

Simplified the slice copy/update logic.  Removed the panics for non
required interface functions, the impl was trivial, added simple tests
to ensure they work as expected.
